nibabel/tests/test_image_api.py

```diff
@@ -451,6 +452,73 @@ def validate_affine_deprecated(self, imaker, params):
             assert_true(aff is img.get_affine())
 
 
+class SerializeMixin(object):
+    def validate_to_bytes(self, imaker, params):
+        img = imaker()
+        serialized = img.to_bytes()
+        with InTemporaryDirectory():
+            fname = 'img' + self.standard_extension
+            img.to_filename(fname)
+            with open(fname, 'rb') as fobj:
+                file_contents = fobj.read()
+        assert serialized == file_contents
+
+    def validate_from_bytes(self, imaker, params):
+        img = imaker()
+        klass = getattr(self, 'klass', img.__class__)
+        with InTemporaryDirectory():
+            fname = 'img' + self.standard_extension
+            img.to_filename(fname)
+
+            all_images = list(getattr(self, 'example_images', [])) + [{'fname': fname}]
+            for img_params in all_images:
+                img_a = klass.from_filename(img_params['fname'])
+                with open(img_params['fname'], 'rb') as fobj:
+                    img_b = klass.from_bytes(fobj.read())
+
+                assert self._header_eq(img_a.header, img_b.header)
+                assert np.array_equal(img_a.get_data(), img_b.get_data())
+                del img_a
+                del img_b
+
+    def validate_to_from_bytes(self, imaker, params):
+        img = imaker()
+        klass = getattr(self, 'klass', img.__class__)
+        with InTemporaryDirectory():
+            fname = 'img' + self.standard_extension
+            img.to_filename(fname)
+
+            all_images = list(getattr(self, 'example_images', [])) + [{'fname': fname}]
+            for img_params in all_images:
+                img_a = klass.from_filename(img_params['fname'])
+                bytes_a = img_a.to_bytes()
+
+                img_b = klass.from_bytes(bytes_a)
+
+                assert img_b.to_bytes() == bytes_a
+                assert self._header_eq(img_a.header, img_b.header)
+                assert np.array_equal(img_a.get_data(), img_b.get_data())
+                del img_a
+                del img_b
+
+    @staticmethod
+    def _header_eq(header_a, header_b):
+        """ Quick-and-dirty header equality check
+
+        Abstract classes may have undefined equality, in which case test for
+        same type
+        """
+        not_implemented = False
+        header_eq = True
+        try:
+            header_eq = header_a == header_b
+        except NotImplementedError:
+            header_eq = type(header_a) == type(header_b)
```

Comment:
Should we require a header equality check being defined?  This doesn't seem like a good test.

Reply:
This is a hack to compare `FileBasedHeader`s. I can put in a specific carve-out for that type, instead.